File: scripts/migrate-gen-imports.mjs
import fs from 'node:fs/promises'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const scriptDir = path.dirname(fileURLToPath(import.meta.url))
const pkgDir = path.resolve(scriptDir, '..', 'pkg')

const importPathPrefix = 'github.com/qomos-w/sporemind/pkg/domain/gen/'
const genImportLine = '\tgen "github.com/qomos-w/sporemind/pkg/domain/gen"'

async function main() {
  const files = await walk(pkgDir)
  for (const file of files) {
    if (!file.endsWith('.go')) continue
    const src = await fs.readFile(file, 'utf8')
    const result = transform(src)
    if (result.changed) {
      await fs.writeFile(file, result.source, 'utf8')
      console.log(`[migrate-gen-imports] ${path.relative(pkgDir, file)}`)
    }
  }
}

function transform(src) {
  // Normalise CRLF to LF for easier regex handling; write back LF.
  const normalizedSrc = src.replace(/\r\n/g, '\n')
  let changed = false
  let newSrc = normalizedSrc

  // Identify every local alias that points to a domain/gen/* import.
  const aliasToPath = new Map() // alias -> full import path

  const importLineRegex = new RegExp(`^\\s*([a-zA-Z_][a-zA-Z0-9_]*)?\\s*"${importPathPrefix}([^"]+)"`, 'gm')
  let m
  while ((m = importLineRegex.exec(normalizedSrc)) !== null) {
    const alias = m[1] || m[2]
    aliasToPath.set(alias, `${importPathPrefix}${m[2]}`)
  }

  if (aliasToPath.size === 0) {
    return { source: src, changed: false }
  }
  changed = true

  // Replace every qualified identifier that uses one of those aliases.
  for (const alias of aliasToPath.keys()) {
    newSrc = newSrc.replace(new RegExp(`\\b${alias}\\.`, 'g'), 'gen.')
  }

  // Rebuild import block(s).
  newSrc = newSrc.replace(/import\s*\(\r?\n([\s\S]*?)\r?\n\)/g, (match, body) => {
    const lines = body.split('\n')
    const kept = []
    let hasGen = false

    for (const rawLine of lines) {
      const line = rawLine.replace(/\r/g, '')
      const trimmed = line.trim()
      if (trimmed === '' || trimmed.startsWith('//')) {
        kept.push(line)
        continue
      }
      const m = trimmed.match(/^([a-zA-Z_][a-zA-Z0-9_]*)?\s*"([^"]+)"$/)
      if (!m) {
        kept.push(line)
        continue
      }
      const p = m[2]
      if (p.startsWith(importPathPrefix)) {
        // Drop old gen subpackage import.
        continue
      }
      if (p === 'github.com/qomos-w/sporemind/pkg/domain/gen') {
        hasGen = true
      }
      kept.push(line)
    }

    if (!hasGen) {
      kept.push(genImportLine)
    }

    const sorted = kept
      .filter(l => l.trim() !== '')
      .sort((a, b) => {
        const at = a.trim().replace(/^\w+\s+/, '"')
        const bt = b.trim().replace(/^\w+\s+/, '"')
        return at.localeCompare(bt)
      })

    const normalized = sorted.map(l => l.startsWith('\t') ? l : '\t' + l.trim())
    return `import (\n${normalized.join('\n')}\n)`
  })

  // Handle single-line import statements.
  newSrc = newSrc.replace(/import\s+([a-zA-Z_][a-zA-Z0-9_]*)?\s*"([^"]+)"\n/g, (match, alias, p) => {
    if (!p.startsWith(importPathPrefix)) return match
    return `import gen "github.com/qomos-w/sporemind/pkg/domain/gen"\n`
  })

  return { source: newSrc, changed }
}

async function walk(dir) {
  const entries = await fs.readdir(dir, { withFileTypes: true })
  const files = []
  for (const entry of entries) {
    const full = path.join(dir, entry.name)
    if (entry.isDirectory()) {
      files.push(...await walk(full))
    } else {
      files.push(full)
    }
  }
  return files
}

main().catch(err => {
  console.error('[migrate-gen-imports] failed:', err)
  process.exitCode = 1
})

File: scripts/migrate-gen-imports.go
// Command migrate-gen-imports rewrites Go files so that imports of
// github.com/qomos-w/sporemind/pkg/domain/gen/<subpkg> become a single
// gen "github.com/qomos-w/sporemind/pkg/domain/gen" import and all
// qualified references to those subpackages become gen.<Name>.
//
// It is AST-aware: field accesses on local variables whose names happen to
// match an imported package name are left untouched.
package main

import (
	"fmt"
	"go/ast"
	"go/format"
	"go/parser"
	"go/token"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

const genPath = "github.com/qomos-w/sporemind/pkg/domain/gen"
const genPathPrefix = genPath + "/"

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintln(os.Stderr, "usage: migrate-gen-imports <dir>")
		os.Exit(2)
	}
	root := os.Args[1]

	var files []string
	err := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if !d.IsDir() && strings.HasSuffix(path, ".go") {
			files = append(files, path)
		}
		return nil
	})
	if err != nil {
		fmt.Fprintln(os.Stderr, "walk:", err)
		os.Exit(1)
	}

	sort.Strings(files)
	for _, f := range files {
		changed, err := rewriteFile(f)
		if err != nil {
			fmt.Fprintf(os.Stderr, "%s: %v\n", f, err)
			os.Exit(1)
		}
		if changed {
			rel, _ := filepath.Rel(root, f)
			fmt.Println(rel)
		}
	}
}

func rewriteFile(path string) (bool, error) {
	fset := token.NewFileSet()
	f, err := parser.ParseFile(fset, path, nil, parser.ParseComments)
	if err != nil {
		return false, fmt.Errorf("parse: %w", err)
	}

	// Determine which import names refer to old gen subpackages.
	oldNames := make(map[string]bool)
	for _, imp := range f.Imports {
		p, err := strconvUnquote(imp.Path.Value)
		if err != nil {
			continue
		}
		if !strings.HasPrefix(p, genPathPrefix) {
			continue
		}
		name := importName(imp, p)
		if name == "_" || name == "." {
			continue
		}
		oldNames[name] = true
	}
	if len(oldNames) == 0 {
		return false, nil
	}

	// Rewrite selector expressions: oldName.Ident -> gen.Ident
	ast.Inspect(f, func(n ast.Node) bool {
		sel, ok := n.(*ast.SelectorExpr)
		if !ok {
			return true
		}
		ident, ok := sel.X.(*ast.Ident)
		if !ok {
			return true
		}
		if !oldNames[ident.Name] {
			return true
		}
		ident.Name = "gen"
		return true
	})

	// Rebuild imports: drop all gen subpackage imports, ensure gen import exists.
	var newImports []*ast.ImportSpec
	hasGen := false
	for _, imp := range f.Imports {
		p, err := strconvUnquote(imp.Path.Value)
		if err != nil {
			newImports = append(newImports, imp)
			continue
		}
		if strings.HasPrefix(p, genPathPrefix) {
			continue
		}
		if p == genPath {
			hasGen = true
			if imp.Name == nil || imp.Name.Name == "" {
				imp.Name = &ast.Ident{Name: "gen"}
			}
		}
		newImports = append(newImports, imp)
	}
	if !hasGen {
		newImports = append(newImports, &ast.ImportSpec{
			Name: &ast.Ident{Name: "gen"},
			Path: &ast.BasicLit{Kind: token.STRING, Value: `"` + genPath + `"`},
		})
	}

	// Sort imports by import path (same order gofmt uses).
	sort.SliceStable(newImports, func(i, j int) bool {
		pi, _ := strconvUnquote(newImports[i].Path.Value)
		pj, _ := strconvUnquote(newImports[j].Path.Value)
		return pi < pj
	})

	f.Imports = newImports

	// Replace the import declaration node in the file.
	for i, decl := range f.Decls {
		genDecl, ok := decl.(*ast.GenDecl)
		if !ok || genDecl.Tok != token.IMPORT {
			continue
		}
		genDecl.Specs = nil
		for _, imp := range newImports {
			genDecl.Specs = append(genDecl.Specs, imp)
		}
		if len(genDecl.Specs) == 0 {
			f.Decls = append(f.Decls[:i], f.Decls[i+1:]...)
		}
		break
	}

	var buf strings.Builder
	if err := format.Node(&buf, fset, f); err != nil {
		return false, fmt.Errorf("format: %w", err)
	}
	out := buf.String()

	// Preserve original line endings (CRLF vs LF) by matching input.
	srcBytes, err := os.ReadFile(path)
	if err != nil {
		return false, fmt.Errorf("read: %w", err)
	}
	if strings.Contains(string(srcBytes), "\r\n") {
		out = strings.ReplaceAll(out, "\n", "\r\n")
	}

	if out == string(srcBytes) {
		return false, nil
	}
	if err := os.WriteFile(path, []byte(out), 0o644); err != nil {
		return false, fmt.Errorf("write: %w", err)
	}
	return true, nil
}

func strconvUnquote(s string) (string, error) {
	return strings.Trim(s, `"`), nil
}

func importName(imp *ast.ImportSpec, path string) string {
	if imp.Name != nil && imp.Name.Name != "" {
		return imp.Name.Name
	}
	i := strings.LastIndex(path, "/")
	if i >= 0 {
		return path[i+1:]
	}
	return path
}
